### PR TITLE
Add more repo search tests and /api/repo/search integration tests 

### LINKED ISSUE #2446: Integration test TestAPISearchRepoNotLogin doesn't work
Integration test TestAPISearchRepoNotLogin doesn't work. Cuased by #2125 (custom struct `searchResponseBody` has unexported values that can't be use in json decode function)

### PR DESCRIPTION
**Pull request by Morlinest**
Add some more tests. Also fixes #2446.

Two integration tests will not pass with current implementation of repo search api because it doesn't return all repositories accessible to user if he is not logged or is different from logged user

### CODE REVIEW

**Comment by lunny** [comment]
CI failed.

**Comment by Morlinest** [comment]
@lunny I told you it will and why.

I've added 8 new repositories to tests, 4 (2x public, 2x private) are owned by user15, 2 (1x public, 1x private) owned by user16 and user15 is collaborator in both of them and 2 repositories (1x public, 1x private) owned by organization which is owned by user15. That means if you search for public repositories accessible (and related) to user15, you should get 4 public repositories (2 owned by user15, 1 owned by user16 and 1 owned by organization he owns).

You should get only public (directly owned or accessible + related by any other non-private rules) repositories of user if it is anonymous search or logged user is different to that user. Am I thinking right?

File: models/repo_list_test.go [hunk 1]
```diff
@@ -60,10 +57,109 @@ func TestSearchRepositoryByName(t *testing.T) {
...
+	repos, count, err = SearchRepositoryByName(&SearchRepoOptions{
+		Page:     1,
+		PageSize: 10,
+		OwnerID:  15,
+		Searcher: &User{ID: 15},
+	})
+
+	assert.NotNil(t, repos)
+	assert.NoError(t, err)
+	assert.Equal(t, int64(3), count)
```
Comment (Morlinest) [inline models/repo_list_test.go]: @lunny @lafriks I think, this should return 2, not 3 (only directly owned repositories). Organization repositories should be included with `collaborate` flag. What do you think? Same applies to line 112. Or add something like `organization` flag.

Comment (lunny) [reply models/repo_list_test.go]: `Orgnization` repositories should be in `collaborate`, I think it's right.

**Comment by Morlinest** [comment]
Fixed tests and added 2 new tests for pages and pagesize limit

**Comment by Morlinest** [comment]
Tests failed as expected with current implementation of api and repo search function. I've done tests with #2371 and it will pass these new tests (#2371 rebased on this PR can be found in [test-repo-search-implementation](https://github.com/Morlinest/gitea/tree/test-repo-search-implementation)).

**Comment by Morlinest** [comment]
Merged to #2371.

**Pull request closed without merging**